Integer determinant calculation

insalt-glitch opened the pull request:
Resolves #73 . I added an overload for the determinant calculation with [Bareiss algorithm](https://en.wikipedia.org/wiki/Bareiss_algorithm). It is probably slightly slower than the `LUDecomposition` but is guaranteed to be correct. I hope this fine :).

tomstewart89 commented:
Great work! Unfortunately if I compile this using the arduino IDE I get:
```
BasicLinearAlgebra/impl/NotSoBasicLinearAlgebra.h:327:15: error: 'enable_if' in namespace 'std' does not name a template type
 typename std::enable_if<std::is_integral<typename ParentType::DType>::value, typename ParentType::DType>::type
               ^~~~~~~~~
```
Offhand there might be some way we can achieve the same effect with `static_asserts` somehow? Failing that, if this implementation supports integer datatypes and works well then I'm happy to just drop the current implementation in favor of this one. Let me know how you go! Thanks for the PR!

tomstewart89 commented on impl/NotSoBasicLinearAlgebra.h:
```diff
@@ -322,6 +323,46 @@ typename ParentType::DType Determinant(const MatrixBase<ParentType, Dim, Dim, ty
+                if (A_copy(idx, i) != 0) break;
+            }
+            if (idx == Dim) return 0;
+            const BLA::Matrix<1, Dim, typename ParentType::DType> tmp = A_copy.Row(i);
```
```suggestion
            const auto tmp = A_copy.Row(i);
```
Row will return a `RefMatrix` which doesn't copy the original array. Since you're not modifying tmp then it's probably a bit more efficient to avoid taking a copy here

insalt-glitch replied:
Sorry, I'm a bit confused, and the code is too. We are trying to swap the two rows, I think for that we need a temporary copy. I cannot use the `std::swap` or `bla_swap` methods because the assignment operator is deleted for `RefMatrix`. Do I have something wrong?

Also, I'm not sure if the `auto` will make a `RefMatrix` or `Matrix` type.

tomstewart89 replied:
Ahh my mistake! Sorry, I should have read this more closely. No problems, it makes sense to take a copy if you're swapping rows :+1:

insalt-glitch commented:
Okay... two things:
- I fixed the compiler-error on for the Arduino target. It should compile now, if you check again. The only problem was that I did not add the `type_traits` header from the standard library. I also simplified the template definition a little bit, such that things are more readable.
- I added a swap-function because it is quite annoying that you cannot swap matrices at all. Maybe this should end up in another pull-request? Your decision.

Now, I changed the name of the two swap-functions that exist now, because they were not in line with the snake-case naming of the rest of the repository. Should we keep the original names for backwards compatibility?

Additionally, I'm unsure if we should keep the original swap implementation at all. It is basically a simplified version of [std::swap](https://en.cppreference.com/w/cpp/algorithm/swap) in the standard library, and it is only used once in the entire repository. Again, your decision!

Let me know what you think :)

Incorrectone commented:
I do not think the library type_traits is ported to some arduino boards and you will have to use dependencies for this.

1. [Forum post about type_traits](https://forum.arduino.cc/t/compilation-error-type-traits-library-not-found-when-using-umate-and-serial9b-libraries/1146309)
2. [Possible dependency if you want to use them instead](https://www.arduino.cc/reference/en/libraries/arxtypetraits/)

I tried pull library(idk how to say it) on Arduino Mega in Arduino IDE 2 and shows that it cannot find the <type_traits> library.

```
In file included from ***\Arduino\libraries\BasicLinearAlgebra/BasicLinearAlgebra.h:181:0,
                 from ***:
***\Arduino\libraries\BasicLinearAlgebra/impl/NotSoBasicLinearAlgebra.h:3:10: fatal error: type_traits: No such file or directory
 #include <type_traits>
          ^~~~~~~~~~~~~
compilation terminated.
exit status 1

Compilation error: exit status 1
```

Incorrectone commented:
So I do not know much about classes and coding practices, but can we not simply overload the Determinant function for taking some integer type?

`typename ParentType::DType Determinant(const MatrixBase<ParentType, Dim, Dim, int> &A)` without ever using type_traits?

As I tried this with insalt's integer [Bareiss algorithm](https://en.wikipedia.org/wiki/Bareiss_algorithm):

in NotSoBasicLinearAlgebra.h
```cpp
// general case for floats and doubles
template <typename ParentType, int Dim>
typename ParentType::DType Determinant(const MatrixBase<ParentType, Dim, Dim, typename ParentType::DType> &A)
{
    Matrix<Dim, Dim, typename ParentType::DType> A_copy = A;

    auto decomp = LUDecompose(A_copy);

    typename ParentType::DType det = decomp.parity;

    for (int i = 0; i < Dim; ++i)
    {
        det *= decomp.U(i, i);
    }

    return det;
}

// Function specifically for ints

template <typename ParentType, int Dim>
typename ParentType::DType Determinant(const MatrixBase<ParentType, Dim, Dim, int> &A)
{
    // For integral types use Bareiss algorithm from insalt's repo
    Matrix<Dim, Dim, typename ParentType::DType> A_copy = A;

    int sign = 1;
    typename ParentType::DType prev = 1;

    for (int i = 0; i < Dim; i++)
    {
        if (A_copy(i, i) == 0)
        {
            int idx = i + 1;
            for (; idx < Dim; idx++)
            {
                if (A_copy(idx, i) != 0) break;
            }
            if (idx == Dim) return 0;
            const BLA::Matrix<1, Dim, typename ParentType::DType> tmp = A_copy.Row(i);

            for (int k = 0; k < Dim; k++)
            {
                A_copy(i, k) = A_copy(idx, k);
                A_copy(idx, k) = tmp(1, k);
            }
            sign = - sign;
        }
        for (int j = i + 1; j < Dim; j++)
        {
            for (int k = i + 1; k < Dim; k++)
            {
                A_copy(j, k) = (A_copy(j, k) * A_copy(i, i) - A_copy(j, i) * A_copy(i, k)) / prev;
            }
        }
        prev = A_copy(i, i);
    }
    
    return sign * A_copy(Dim - 1, Dim - 1);
}
```

Also one more thing for type conversion from float matrices to int and vice-versa

Can we add
in ElementStorage.h
```cpp
    template <typename CastType> 
    Matrix(const Matrix<Rows, Cols, CastType> &other){ 
        for (int i = 0; i < Rows; ++i)
        {
            for (int j = 0; j < Cols; ++j)
            {
                (*this)(i, j) = static_cast<DType>(other(i,j));
            }
        }
    }
```

This also gives the correct answer for integer matrices

insalt-glitch commented:
> I do not think the library type_traits is ported to some Arduino boards, and you will have to use dependencies for this.
> 
>     1. Forum post about type_traits: [1](https://forum.arduino.cc/t/compilation-error-type-traits-library-not-found-when-using-umate-and-serial9b-libraries/1146309) [2](https://stackoverflow.com/questions/37575303/is-the-c-standard-library-fully-supported-on-arduino)
> 
>     2. [Possible dependency if you want to use them instead](https://www.arduino.cc/reference/en/libraries/arxtypetraits/) (Edit: I have not tried this yet, also it does not have the is_fundamental, it does however have is_same)
> 
>     3. Or one can always use their own implemented functions instead if you want no dependencies
> 
> 
> I tried the pull library (idk how to say it) on Arduino Mega in Arduino IDE 2, and it shows that it cannot find the <type_traits> library.
> 
> ```
> In file included from ***\Arduino\libraries\BasicLinearAlgebra/BasicLinearAlgebra.h:181:0,
>                  from ***:
> ***\Arduino\libraries\BasicLinearAlgebra/impl/NotSoBasicLinearAlgebra.h:3:10: fatal error: type_traits: No such file or directory
>  #include <type_traits>
>           ^~~~~~~~~~~~~
> compilation terminated.
> exit status 1
> 
> Compilation error: exit status 1
> ```

Thank you very much for investigating. I only tried with the`Arduino MKR Zero` as a board and there it is apparently implemented. I will rewrite the implementation to remove the dependency. I also think, at least for me personally, a huge charm of this library is the fact that it does not have any other library dependencies. So I'll come up with a strategy on the weekend and present may new idea.

The template specialization for `int` does not sound very appealing to me, since it would only work for `int32_t`, what about `int16_t`, etc. I guess we should decide what to do about unsigned types though. There, the current implementation gives the wrong answer. Probably in these cases we should always return an `int`, or emit an error with `static_assert`? I'm open for suggestions.

insalt-glitch commented:
Okay, I changed the implementation now such that the type-traits header is no longer included, and I hope everything should compile fine now. The question still remains whether we want the additional complexity that this changes brings, and if anyone has an idea to make it simpler.

tomstewart89 commented on impl/NotSoBasicLinearAlgebra.h:
```diff
@@ -135,14 +143,19 @@ LUDecomposition<ParentType> LUDecompose(MatrixBase<ParentType, Dim, Dim, typenam
             decomp.parity = -decomp.parity;
 
-            bla_swap(idx[j], idx[argmax]);
+            // swap indicies
```
```suggestion
            // swap indices
```
nit: funnily enough, I often make this exact same spelling error

tomstewart89 commented on impl/Types.h:
```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+namespace BLA
```
This file looks great. I agree it adds a bit of complexity to this library but it's really nice to be able to reason abut the DType of the Matrix when writing linear algebra functions. Great work!

tomstewart89 reviewed (approved):
This PR looks great! I agree that it brings a little bit of extra complexity to this library but I'm totally with it and I think the `Types.h` functions are a handy addition to this library. 

Thanks for all your work on this one!
